File: app/src/main/java/com/burhan/arch/room/models/DiffUtilUser.java
package com.burhan.arch.room.models;

import android.support.v7.util.DiffUtil;

import java.util.List;

/**
 * Created by Rashid on 10/4/2017.
 */

public class DiffUtilUser extends DiffUtil.Callback {

    private List<User> newUserList;
    private List<User> oldUserList;


    public DiffUtilUser(List<User> newUserList, List<User> oldUserList) {
        this.newUserList = newUserList;
        this.oldUserList = oldUserList;
    }

    @Override
    public int getOldListSize() {
        return oldUserList != null ? oldUserList.size() : 0;
    }

    @Override
    public int getNewListSize() {
        return newUserList != null ? newUserList.size() : 0;
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldUserList.get(oldItemPosition).getUid() == newUserList.get(newItemPosition).getUid();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return oldUserList.get(oldItemPosition).equals(newUserList.get(newItemPosition));
    }
}

File: app/src/main/java/com/burhan/arch/room/models/DiffUtilUser.kt
package com.burhan.arch.room.models

import androidx.recyclerview.widget.DiffUtil

/**
 * Created by Rashid on 10/4/2017.
 */
class DiffUtilUser(private val newUserList: List<User>, private val oldUserList: List<User>) : DiffUtil.Callback() {
    override fun getOldListSize(): Int {
        return oldUserList.size
    }

    override fun getNewListSize(): Int {
        return newUserList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldUserList[oldItemPosition].uid == newUserList[newItemPosition].uid
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldUserList[oldItemPosition].equals(newUserList[newItemPosition])
    }

}